From 733cab75650a8ab6e10d16c4583c8ab711ec1056 Mon Sep 17 00:00:00 2001
From: Raffaele Aquilone <raquiloX>
Date: Mon, 24 Jun 2013 13:47:59 +0200
Subject: [PORT FROM MAIN][BT] Improve random BD address generation

BZ: 116152

Using gettimeofday for reading current time in order to have a better seed
to initialize the pseudorandom number generator, used to assign a random
BD address to the BT chip when the CHAABI doesn't provide one

Change-Id: I0c2d79bf337307ab442b0a5a0a1df16fc91efa96
Orig-Change-Id: Id7d0287199a8a81b5d6538cf4c3f07cb11761322
Category: device enablement
Domain: CWS.BT-Common
Origin: internal
Upstream-Candidate: yes
Signed-off-by: Raffaele Aquilone <raquiloX>
Reviewed-by: Gros, Julien <julien.gros@intel.com>
Reviewed-by: Poignet, Cedric <cedric.poignet@intel.com>
Reviewed-on: http://android.intel.com:8080/122991
Tested-by: Champciaux, Nicolas <nicolas.champciaux@intel.com>
---
 btif/src/btif_core.c |   13 ++++++++++++-
 1 files changed, 12 insertions(+), 1 deletions(-)

diff --git a/btif/src/btif_core.c b/btif/src/btif_core.c
index 40f19d0..40857f0 100644
--- a/btif/src/btif_core.c
+++ b/btif/src/btif_core.c
@@ -30,6 +30,7 @@
 #include <string.h>
 #include <sys/types.h>
 #include <sys/stat.h>
+#include <sys/time.h>
 #include <fcntl.h>
 #include <dirent.h>
 #include <ctype.h>
@@ -417,8 +418,18 @@ static void btif_fetch_local_bdaddr(bt_bdaddr_t *local_addr)
     if (!valid_bda)
     {
         bdstr_t bdstr;
+        struct timeval tval;
         /* Seed the random number generator */
-        srand((unsigned int) (time(0)));
+        int ret = gettimeofday(&tval, NULL);
+        if (!ret) {
+            BTIF_TRACE_DEBUG2("gettimeofday returns %d secs and %d usecs",
+                                                    tval.tv_sec, tval.tv_usec);
+            srand((unsigned int) tval.tv_sec * tval.tv_usec);
+        } else {
+            BTIF_TRACE_WARNING1("gettimeofday failed with error %s",
+                                                            strerror(errno));
+            srand((unsigned int) time(NULL));
+        }
 
         /* No autogen BDA. Generate one now. */
         local_addr->address[0] = 0x22;
-- 
1.7.4.1

